### ISSUE TITLE
Issues with syncing with TinyTinyRSS

### ISSUE THREAD

**Issue by The-Compiler** [newsboat/newsboat#16]
While I love Newsb{oat,euter}, syncing updates with my TinyTinyRSS server always has been a lot of pain. A quick overview of what I've noticed so far, in the past few months to years (note that I'm often in a train, i.e. on an unstable connection):

- Rendering issues (random chars, articles disappearing, ...) when viewing an article while updating (though this kind of thing has gotten better recently-ish)
- Crashes (C++ exceptions) when doing the same (dito)
- A lot of HTTP requests being done while syncing, and syncing being very slow on a bad network
- Articles getting marked as unread again after having read them
- The most frustrating one: Articles randomly getting marked as read again after "peeking" at them and marking them as unread with `N`.

The first two things have gotten better, and I realize I should submit a better bug report when I see them again - I know this isn't :wink:

As for the rest - I wonder if this just happens because of newsbeuter's sync "model"? I guess when I do `N`, it immediately sends a request to the server to mark it as unread again, and if that request gets lost, what I mentioned happens?

This has made me wonder: Maybe it'd be better to have some kind of explicit syncing, similar to what e.g. [offlineimap](http://www.offlineimap.org/) does for mails? In other words, I could hit some key to update, then everything happens locally, and when I hit that key again, it tries to synchronize whatever has happened to the server?

If this is infeasible in Newsbeuter itself, I was wondering whether you think it'd make sense to write this as a separate utility, i.e. something that talks to TinyTinyRSS and updates newsbeuter's local files somehow.

I know this isn't really an actionable issue, but I'd love to hear your thoughts on it!

**Comment by Minoru** [comment]
> Rendering issues (random chars, articles disappearing, ...) when viewing an article while updating (though this kind of thing has gotten better recently-ish)

This is a general problem, not specific to remote APIs. Seems like it can be solved by changing some low-levels bits that hold feeds and items together; I keep postponing a dive into that part of the code. Maybe after the cache revamp?

> Crashes (C++ exceptions) when doing the same (dito)

That's new! Will continue in #17.

The rest, I think, are indeed caused by our simplistic synchronization model. Newsboat doesn't even retry failed requests!

Ideally Newsboat would immediately apply user's actions to local DB *and* enqueue them to be sent to remote server. The sending itself will happen in a separate thread, and queue will be saved/restored when user quits Newsboat.

Manual syncing seems like a nice first step towards that ideal. Don't see any reason to make it a separate program—I do want this code as part of Newsboat itself.

This will most probably require a rework of our remote api interface (`include/remote_api.h`), but I'm ready to help out in any way I can. Seems like the only thing we're missing is a volunteer who will do the work. Maybe you're the one? :)

**Comment by The-Compiler** [comment]
> The rest, I think, are indeed caused by our simplistic synchronization model. Newsboat doesn't even retry failed requests!

Yeah, that explains it then! I wonder, what do you think about something like a `peek` command which opens an article but doesn't mark it as read?

> Seems like the only thing we're missing is a volunteer who will do the work. Maybe you're the one? :)

I know that feeling! Probably not anytime soon I'm afraid - I'm more than busy with [my own project](https://www.qutebrowser.org/).

**Comment by Minoru** [comment]
> I wonder, what do you think about something like a peek command which opens an article but doesn't mark it as read?

Sorry, I thought I answered this two weeks ago! Obviously I somehow didn't.

My view on such a command is two-fold:

1. if you're thinking of it just to work around the problem with unread status synchronization, I'd rather work on fixing the synchronization itself.

2. if you have a use-case for it, please describe it. Why can't it be implemented with a macro (`open ; toggle-unread` should do it)? What happens if I peek into an article and go to the next one without quitting the article view?

**Comment by noctux** [comment]
Getting back to the original problem:
The sync model of newsboat<->TTRSS is no real sync model. The flow is more like this: When you refresh your feedlist from TTRSS, currently newsboat will download the latest X (for X \in {60, 200}, depending on your TTRSS version, with 200 for newer versions) articles per feed including the full body using `getHeadlines` and then merge the results. Depending on the server side and your connection, there is certainly room for improvement and we have a rather slow sync.

Ideas that come to mind to alleviate the problem:
Since version 1.5.6 of the [upstream api](https://git.tt-rss.org/git/tt-rss/wiki/ApiReference), there is a `since_id` operand that requests only articles that were added later than the most recent we have synced. But if we go down that road, synchronisation of read state as well as content updates for older articles have to be reconsidered.

Another note: The way we currently detect upstream feeds is also suboptimal. We currently enumerate all categories, and then request the feeds for each category. Since api-level 5, there is an `getFeedTree` api function that might allow us to do the whole thing in just one request instead.

I still intend to look into both issues one day, but I cannot give any guarantees :)

To add to the problem: Newsboats current sync model for read-state means that updates to the read state of articles might easily get lost. Whenever we read an article, we simply issue a short api-call to mark the resulting article as read. However, if this fails (e.g. because we are offline at the moment, ...), no further action is taken and read/write state gets out of sync. As far as I can tell, this holds for all remote apis. When discussing the design of the remote api, this should be considered as well.

Finally: Currently, we do funny things like constructing json payloads by string concatenation, which will fail in cases like usernames/passwords containing quotationmarks, etc.pp.. For newsbeuter, there has been an issue discussing switching to a "better" (for sufficient values of better) json library. When a decision is made there, we should consider using that library for constructing our json-payloads.

Ok, this got a bit longer than intended, but I hope this helps.
